File: Python/src/sorts/radix_sort.py
def radix_sort(array):
    max = array[0]
    for i in range(1, len(array)):
        if array[i] > max:
            max = array[i]

    exp = 1
    while max // exp > 0:
        _count_sort(array, exp)
        exp *= 10

def _count_sort(array, exp):
    n = len(array)
    output = [0] * n
    count = [0] * 10

    for i in range(0, n):
        digit = (array[i] // exp) % 10
        count[digit] += 1
    
    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        digit = (array[i] // exp) % 10
        output[count[digit] - 1] = array[i]
        count[digit] -= 1

    for i in range(0, n):
        array[i] = output[i]

File: CSharp/src/Sorts/RadixSort.cs
namespace Algorithm.Sorts;

public class RadixSort
{
    public void Sort(int[] array)
    {
        var max = GetMax(array);

        for (var exp = 1; max / exp > 0; exp *= 10)
            CountingSort(array, exp);
    }

    private int GetMax(int[] array)
    {
        var max = array[0];
        for (var i = 1; i < array.Length; i++)
            if (array[i] > max)
                max = array[i];
        return max;
    }

    private void CountingSort(int[] array, int exp)
    {
        var n = array.Length;
        var output = new int[n];
        var count = new int[10];

        for (var i = 0; i < n; i++)
        {
            var digit = (array[i] / exp) % 10;
            count[digit]++;
        }

        for (var i = 1; i < 10; i++)
        {
            count[i] += count[i - 1];
        }

        for (var i = n - 1; i >= 0; i--)
        {
            var digit = (array[i] / exp) % 10;
            output[count[digit] - 1] = array[i];
            count[digit]--;
        }

        for (var i = 0; i < n; i++)
        {
            array[i] = output[i];
        }
    }
}